shap/explainers/tree.py

```diff
@@ -160,24 +164,69 @@ def shap_values(self, X, **kwargs):
             else:
                 return [phi[:, :, i] for i in range(self.n_outputs)]
 
-    def shap_interaction_values(self, X, **kwargs):
+    def shap_interaction_values(self, X, tree_limit=-1, **kwargs):
 
         # shortcut using the C++ version of Tree SHAP in XGBoost and LightGBM
         if self.model_type == "xgboost":
             if not str(type(X)).endswith("xgboost.core.DMatrix'>"):
                 X = xgboost.DMatrix(X)
-            phi = self.trees.predict(X, pred_interactions=True)
+            if tree_limit==-1:
+                tree_limit=0
+            phi = self.trees.predict(X, ntree_limit=tree_limit, pred_interactions=True)
             if len(phi.shape) == 4:
                 return [phi[:, i, :, :] for i in range(phi.shape[1])]
             else:
                 return phi
         else:
-            raise Exception("Interaction values not yet supported for model type: " + self.model_type)
+            if str(type(X)).endswith("pandas.core.series.Series'>"):
+                X = X.values
+            elif str(type(X)).endswith("pandas.core.frame.DataFrame'>"):
+                X = X.values
+
+            assert str(type(X)).endswith("'numpy.ndarray'>"), "Unknown instance type: " + str(type(X))
+            assert len(X.shape) == 2, "Instance must have 2 dimensions!"
+
+            self.n_outputs = self.trees[0].values.shape[1]
+
+            x_missing = np.zeros(X.shape[1], dtype=np.bool)
+            pool = multiprocessing.Pool()
+            self._current_X = X
+            self._current_x_missing = x_missing
+
+            # Only python 3 can serialize a method to send to another process
+            if sys.version_info[0] >= 3:
+                phi = np.stack(pool.map(self._tree_shap_ind_interactions, range(X.shape[0])), 0)
+            else:
+                phi = np.stack(map(self._tree_shap_ind_interactions, range(X.shape[0])), 0)
+
+            if self.n_outputs == 1:
+                return phi[:, :, :, 0]
+            else:
+                return [phi[:, :, :, i] for i in range(self.n_outputs)]
 
-    def _tree_shap_ind(self, i):
+    def _tree_shap_ind(self, i, condition=0, condition_feature=0):
         phi = np.zeros((self._current_X.shape[1] + 1, self.n_outputs))
         for t in self.trees:
-            self.tree_shap(t, self._current_X[i,:], self._current_x_missing, phi)
+            self.tree_shap(t, self._current_X[i,:], self._current_x_missing, phi, condition, condition_feature)
+        phi /= len(self.trees)
+        return phi
+
+    def _tree_shap_ind_interactions(self, i, condition=0, condition_feature=0):
+        phi = np.zeros((self._current_X.shape[1] + 1, self._current_X.shape[1] + 1, self.n_outputs))
+        phi_diag = np.zeros((self._current_X.shape[1] + 1, self.n_outputs))
+        for t in self.trees:
+            self.tree_shap(t, self._current_X[i,:], self._current_x_missing, phi_diag, condition, condition_feature)
+            unique_features = np.unique(t.features)
```

Comment:
seems like this should happen in the tree constructor not here where it happens again and again for every sample